File: Golang/654. Maximum Binary Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// func constructMaximumBinaryTree(nums []int) *TreeNode { // 108ms 栈
//     if nums == nil {
//         return nil
//     }
//     stack := make([]*TreeNode, 0, len(nums))
//     for _, num := range nums {
//         curr := &TreeNode{Val: num}
//         for len(stack) > 0 && stack[len(stack)-1].Val < num {
//             curr.Left = stack[len(stack)-1]
//             stack = stack[: len(stack)-1]
//         }
//         if len(stack) > 0 {
//             stack[len(stack)-1].Right = curr
//         }
//         stack = append(stack, curr)
//     }
//     return stack[0]
// }

func constructMaximumBinaryTree(nums []int) *TreeNode { // 108ms 递归 速度也不慢
    length := len(nums)
    if length == 0 {
        return nil
    } 
    if length == 1 {
        return &TreeNode{Val: nums[0]}
    }
    max := 0
    index := 0
    for i, num := range nums {
        if num > max {
            max = num
            index = i
        }
    }
    node := &TreeNode{Val: nums[index]}
    node.Left = constructMaximumBinaryTree(nums[: index])
    node.Right = constructMaximumBinaryTree(nums[index+1: ])
    return node
}

File: Python3/654. Maximum Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def constructMaximumBinaryTree(self, nums):
        """
        :type nums: List[int]
        :rtype: TreeNode
        """
        # if len(nums) == 0: # 209ms 递归
        #     return None
        # elif len(nums) == 1:
        #     return TreeNode(nums[0])
        # else:
        #     index = nums.index(max(nums))
        #     node = TreeNode(nums[index])
        #     node.left = self.constructMaximumBinaryTree(nums[: index])
        #     node.right = self.constructMaximumBinaryTree(nums[index+1: ])
        #     return node

        if not nums: # 173ms 栈
            return None
        stack = []
        for num in nums:
            curr = TreeNode(num)
            while stack and stack[-1].val < num:
                curr.left = stack[-1]
                stack.pop()
            if stack:
                stack[-1].right = curr
            stack.append(curr)
        return stack[0]
